tests/test_requests.py

```diff
@@ -550,3 +550,24 @@ async def app(scope, receive, send):
 
     response = client.post("/", data="abc")
     assert response.json() == {"body": "abc", "stream": "abc"}
+
+
+def test_request_middleware_replaces_receive(test_client_factory):
+    """Test the behavior of middleware consuming the request body"""
+
+    async def app(scope: Scope, receive: Receive, send: Send) -> None:
+        # a middleware consumes the body
+        await Request(scope, receive, send).body()
+        # another middleware replaces receive
+
+        async def new_receive() -> Message:
+            return {"type": "http.request", "body": b"data"}
+
+        # the endpoint creates a new Request that uses wrapped_receive
+        data = await Request(scope, new_receive, send).body()
+        await Response(data)(scope, receive, send)
+
+    client = test_client_factory(app)
+    resp = client.get("/", content=b"original data")
+    assert resp.status_code == 200, resp.content
+    assert resp.content == b"data"
```

Comment:
@tomchristie this test passes on master but no longer passes on the `keep-request-state` branch. I'd say this is probably "unsupported" behavior and maybe it's not even worth including as a test, but I wanted to call it out.